io-engine/src/lvs/lvs_lvol.rs

```diff
@@ -243,64 +507,117 @@ impl Share for Lvol {
     fn bdev_uri_original(&self) -> Option<String> {
         None
     }
-}
+}
+
+/// It will give the Common Interface for Volume and Snapshots
+#[async_trait(?Send)]
+pub trait LogicalVolume {
+    /// returns the name of the Logical Volume
+    fn name(&self) -> String;
+
+    /// returns the UUID of the Logical Volume
+    fn uuid(&self) -> String;
+
+    /// return lvs for the Logical Volume
+    fn lvs(&self) -> Lvs;
+
+    /// returns the pool name of the Logical Volume
+    fn pool_name(&self) -> String;
+
+    /// returns the pool uuid of the Logical Volume
+    fn pool_uuid(&self) -> String;
+
+    /// returns a boolean indicating if the Logical Volume is thin provisioned
+    fn is_thin(&self) -> bool;
+
+    /// returns a boolean indicating if the Logical Volume is read-only
+    fn is_read_only(&self) -> bool;
+
+    /// return the size of the Logical Volume in bytes
+    fn size(&self) -> u64;
+
+    /// Returns Lvol disk space usage.
+    fn usage(&self) -> LvolSpaceUsage;
+}
+
+/// It will give Interface Specific to Volume
+#[async_trait(? Send)]
+pub trait VolumeOps: LogicalVolume {
+    // wipe the first 8MB if unmap is not supported on failure the operation
+    // needs to be repeated
+    async fn wipe_super(&self) -> Result<(), Error>;
+
+    /// returns a boolean indicating if the lvol is a snapshot
+    fn is_snapshot(&self) -> bool;
+
+    /// get/read a property from this lvol from disk
+    async fn get(&self, prop: PropName) -> Result<PropValue, Error>;
+
+    /// Create a snapshot
+    async fn create_snapshot(&self, nvmf_req: &NvmfReq, snapshot_name: &str);
```

Comment:
We shouldn't box it unless we need to store this in some kind of struct for example? Maybe caller should be the one to determine.

You can also simply impl this against the actual type, otherwise seems a bit odd to add a trait so you can call a method to get self as another trait?

```suggestion
impl Lvol {
    fn as_lv(&self) -> impl LogicalVolume { self } 
}
```

But do we need this at all since self is already a `LogicalVolume` ?

Reply:
We will have to get rid of NVMe-specific arguments in generic volume operations during implementation of snapshots.